### DISCUSSION TITLE
About LG TV control using BetterDisplay

### DISCUSSION THREAD

**Discussion by SuperNG6** [MonitorControl/MonitorControl#1082] [Ideas]
### Before opening the issue, have you...?

- [X] Searched for existing issues
- [X] Read through [the Monitor-Troubleshooting Wiki](https://github.com/MonitorControl/MonitorControl/wiki/Monitor-Troubleshooting)
- [X] Updated MonitorControl to the latest version (if applicable)

### Describe the issue

Thanks to MonitorControl for this great software and the developers who developed it. I just recently purchased an LG C2 TV for my Mac monitor, and it's realistic, but MonitorControl can't control the LG C2's screen brightness and volume, seemingly because of DDC, which seems to be set up on LG TV's through a network connection

### Expected behavior

Correctly adjust the screen brightness and volume

### Anything else?

no

### Environment Information (please complete the following information)

```markdown
- macOS version:12.4
- Mac model:M1 MacBook Air & Hackintosh(intel 12700,MSI B660M,AMD RX550 LEXA)
- MonitorControl version:4.10
- Monitor(s):LG C2 42
- Monitor Cable(s)/Connection(s): Lenovo HDMI 2.1 Cable
- Apple Silicon/M1 (yes or no): yes
```

**Comment by waydabber** [comment]
Most TVs do not support DDC. They can change their brightness and volume via IR remote only or some proprietary (sometimes open) network control stuff utilized in the TV's remote app. It might be possible to start implementing these in MonitorControl but that is certainly off-scope for now.

I'll convert this one into a discussion, maybe somebody can contribute towards this.

**Comment by waydabber** [comment]
This one is related: https://github.com/MonitorControl/MonitorControl/discussions/1081

**Comment by clipboarder** [comment]
Are there any updates? I am able to control brightness on the LG C2 but not audio, so it's slightly different from the OP. I thought LG OLEDs were supported. Edit: [eqMac](https://eqmac.app/) enables the audio controls. Maybe that's worth looking into.

macOS 13.0.1
MacBook Pro M1 Pro
LG C2 42
HDMI 2.1
Yes

**Reply by waydabber** [reply]
Nope, LG TVs do not support DDC (they support HDMI-CEC for volume control but that is not supported by the Mac). What you should do is turn off DDC control, OLEDs work super well with color table manipulation based brightness control (as there is no backlight). Using eqMac for volume control is a good approach.

**Comment by ShadyNawara** [comment]
hi all, I added network support using bscpylgtv [https://github.com/ShadyNawara/MonitorControl/tree/lg_network](https://github.com/ShadyNawara/MonitorControl/tree/lg_network). I have no intention in maintaining or supporting this in its current form whatsoever but I thought I would share in case somebody really wants to get it working, I might come back to this later and reimplement all of bscpylgtv in swift but until then i dont think its worth the effort when i am just calling shell commands. you would have to atleast update the lgIP under Models->OtherDispaly.swift and potentially the bscpylgtv path under support->LGTVClient.swift.

**Reply by waydabber** [reply]
Hi @ShadyNawara - thanks for this! Sadly I don't have an LG TV to test.

I implemented control integration (using URLs, shell scripts etc) to my other app, BetterDisplay (you'll need the latest v3.5.1 pre-release from here: https://github.com/waydabber/BetterDisplay/releases) - may I ask you (if you have the time of course) to take a look at that and maybe give a feedback if that app works with `bscpylgtv` and LG TV control can be configured fully with it? If it works, it would be great to publish how to set things up properly as I am sure many would like to use their LG TVs with proper control. I am planning to add an option to export/import integration configuration packages and maybe include them in the app for popular options in the future - yours could be a great addition I think.

Let me know if you need any help or want a direct collab on this - you can find me at the app's discord channel: https://discord.gg/aKe5yCWXSp

As an example, here is how I configured my Yamaha AV for volume control, I think setting up a Smart TV for brightness control could be the same (You'll need to use Shell Script integration of course instead of URL):

![image](https://github.com/user-attachments/assets/cea5b4e0-d467-494d-aa4a-a00da0eb4ef0)
![image](https://github.com/user-attachments/assets/f20c8413-c884-4318-9836-63e5ca3be73d)

Once configured properly, the app's default Brightness/Volume sliders should use the integration controls (in this example the volume slider).

![image](https://github.com/user-attachments/assets/dce288d8-e2ba-4cfd-adc4-9d5eab20ff34)

Thank you!

**Reply by ShadyNawara** [reply]
Hi @waydabber thats great thank you for your work on this. I tried to test this new app but i wasnt able to get it working. Is there a way to look at the logs somehow to see what the error could be? Also from your screenshots where do you set the ip address.?

**Reply by waydabber** [reply]
Hi there, you can add parameters.

You should be able to see some logs if you start the app from Terminal and add the -v as argument.

`/Applications/BetterDisplay.app/Contents/MacOS/BetterDisplay -v`

How did you configure things, can you share a screenshot so I may check things out?

**Reply by waydabber** [reply]
This is how a parameter can be added:

https://github.com/waydabber/BetterDisplay/issues/4172

(btw: You can contact me in a DM at Discord or maybe open a discussion here - https://github.com/waydabber/BetterDisplay/discussions)

**Comment by waydabber** [comment]
Note: here is a discord forum about how to configure BetterDisplay with bscpylgtv. After some tweaking it seems to work.

https://discord.com/channels/903276571773771796/1365314093472288860

(discord invite for the app: [discord.gg/aKe5yCWXSp](https://discord.gg/aKe5yCWXSp))

![image](https://github.com/user-attachments/assets/f364f8d2-96e1-466c-9c93-06097a83f40f)

**Comment by ivanfong01** [comment]
I organised the setting for mute and volume for LG TV (tested on LG C5)

PS. i didnt set the brightness as my Mac Studio supports HDR brightness

setting for integration featues:
 - please replace $PATH$ to exact path for python bin location
 - please replace $USERPATH$ to /Users/[you user name]

### Volume control
Sell script (zsh)
`$PATH$/bscpylgtvcommand -p "$USERPATH$/.aiopylgtv.sqlite" [*:IP:*] set_volume [*:VALUE:*]`

Script to get current value (zsh)
`$PATH$/bscpylgtvcommand -p "$USERPATH$/.aiopylgtv.sqlite" [*:IP:*] get_volume`

Return value regex
`(\d+)`


### Mute control
Sell script (zsh) ON
`$PATH$/bscpylgtvcommand -p "$USERPATH$/.aiopylgtv.sqlite" [*:IP:*] set_mute True`

Sell script (zsh) OFF
`$PATH$/bscpylgtvcommand -p "$USERPATH$/.aiopylgtv.sqlite" [*:IP:*] set_mute False`

Script to get current value (zsh)
`$PATH$/bscpylgtvcommand -p "$USERPATH$/.aiopylgtv.sqlite" [*:IP:*] get_muted`

Return content - ON
`True`

Return content - OFF
`False`

**Reply by waydabber** [reply]
Thank you, very useful info! I created a thread about it under BetterDisplay as well and copied this over there so others can get this info under the app discussion as well.

https://github.com/waydabber/BetterDisplay/discussions/4710

**Comment by waydabber** [comment]
Just want to note that I'll add better LG webOS TV support in the next version of BetterDisplay.

<img width="953" height="936" alt="Screenshot 2025-11-11 at 19 33 10" src="https://github.com/user-attachments/assets/7159adfe-6f94-494f-ab04-4f9a5d538d3b" />

https://github.com/waydabber/BetterDisplay/issues/4801

This will include:
- image adjustments
- auto power control (on sleep/wake) for Macs without HDMI-CEC support
- volume control
- virtual remote
- tools to enter service menu easily

**Reply by alperbeser** [reply]
Great! Any ETA for it? 🙂

**Reply by waydabber** [reply]
It is available now if you enable internal pre-release under Settings > Application > Updates.

**Reply by waydabber** [reply]
I am happy to receive some feedback on the feature. In order to make some things work (like power on using WOL), one might have to enable some of the related features on the TV as well.
